File: src/main/java/toker/warbandscripts/panel/service/ChestService.java
package toker.warbandscripts.panel.service;

import org.springframework.stereotype.Service;
import toker.warbandscripts.panel.entity.Chest;
import toker.warbandscripts.panel.entity.Item;
import toker.warbandscripts.panel.entity.pk.ChestPK;
import toker.warbandscripts.panel.repository.ChestRepository;
import toker.warbandscripts.panel.repository.ChestSlotRepository;
import toker.warbandscripts.panel.repository.ItemRepository;

import javax.persistence.criteria.JoinType;
import javax.transaction.Transactional;
import java.util.Optional;

@Service
@Transactional
public class ChestService {

    private ChestRepository chestRepository;
    private ChestSlotRepository chestSlotRepository;
    private ItemRepository itemRepository;

    public ChestService(ChestRepository chestRepository,
                        ChestSlotRepository chestSlotRepository,
                        ItemRepository itemRepository) {
        this.chestRepository = chestRepository;
        this.chestSlotRepository = chestSlotRepository;
        this.itemRepository = itemRepository;
    }

    public Optional<Chest> getChest(int index, int serverId) {
        return chestRepository.findOne(((root, query, builder) -> {
            root.fetch("slots", JoinType.LEFT);
            return builder.and(
                    builder.equal(root.get("index"), index),
                    builder.equal(root.get("server").get("id"), serverId)
            );
        }));
    }

    public void updateChestSlot(int index, int serverId,
                                int slot, int itemId, int ammo) {
        chestSlotRepository.updateChestSlot(chestRepository.getOne(new ChestPK(index, serverId)),
                slot, itemRepository.getOne(itemId), ammo);
    }
}

File: src/main/java/toker/panel/service/ChestService.kt
package toker.panel.service

import org.springframework.stereotype.Service
import toker.panel.entity.Chest
import toker.panel.entity.pk.ChestPK
import toker.panel.repository.ChestRepository
import toker.panel.repository.ChestSlotRepository
import toker.panel.repository.ItemRepository
import java.util.*
import javax.persistence.criteria.CriteriaBuilder
import javax.persistence.criteria.CriteriaQuery
import javax.persistence.criteria.JoinType
import javax.persistence.criteria.Root
import javax.transaction.Transactional

@Service
@Transactional
class ChestService(private val chestRepository: ChestRepository,
                   private val chestSlotRepository: ChestSlotRepository,
                   private val itemRepository: ItemRepository) {
    fun getChest(index: Int, serverId: Int): Optional<Chest> {
        return chestRepository.findOne { root: Root<Chest?>, _, builder: CriteriaBuilder ->
            root.fetch<Any, Any>("slots", JoinType.LEFT)
            builder.and(
                    builder.equal(root.get<Any>("index"), index),
                    builder.equal(root.get<Any>("server").get<Any>("id"), serverId)
            )
        }
    }

    fun updateChestSlot(index: Int, serverId: Int,
                        slot: Int, itemId: Int, ammo: Int) {
        chestSlotRepository.updateChestSlot(chestRepository.getOne(ChestPK(index, serverId)),
                slot, itemRepository.getOne(itemId), ammo)
    }
}